packages/analytics/src/Analytics.ts

```diff
@@ -535,71 +615,82 @@ class Analytics {
   /**
    * Base function to get event data to be sent to integrations.
    *
-   * @param {string} type                           - Type of the event, ex: page, track, onSetUser.
-   * @param {object} [additionalData]               - Additional data to be added to the final event data.
-   * @param {object} [eventContext]                 - Context data that is specific for this event.
+   * @param type - Type of the event, ex: page, track, onSetUser.
+   * @param additionalData  - Additional data to be added to the final event data.
+   * @param eventContext - Context data that is specific for this event.
    *
-   * @returns {Promise<object>}                     - Data for the event.
-   * @private
+   * @returns Data for the event.
    */
-  async getEventData(type, additionalData, eventContext) {
-    const context = await this.context();
+  protected async getEventData<
+    T extends
+      | TrackTypesValues
+      | typeof ON_SET_USER_TRACK_TYPE
+      | typeof LOAD_INTEGRATION_TRACK_TYPE,
+  >(
+    type: T,
+    additionalData?: Partial<EventData<T>>,
+    eventContext?: EventData<T>['context']['event'],
+  ): Promise<EventData<T>> {
+    const context = (await this.context()) as EventContext;
 
     Object.assign(context, { event: eventContext });
 
-    const commonData = {
+    const commonData: Partial<EventData<T>> = {
       type,
-      context,
+      context: context as EventContext,
       platform: this.platform,
-      consent: await this.consent(),
+      consent: (await this.consent()) as ConsentData,
       user: await this.user(),
       timestamp: new Date().getTime(),
     };
 
     return {
       ...commonData,
       ...additionalData,
-    };
+    } as EventData<T>;
   }
 
   /**
    * Returns event data that will be sent on an 'onSetUser' event.
    *
-   * @returns {Promise<object>} - User data to be sent to integrations.
-   * @private
+   * @returns User data to be sent to integrations.
    */
-  async getSetUserEventData() {
-    return await this.getEventData('onSetUser', {
-      event: 'onSetUser',
+  protected async getSetUserEventData(): Promise<SetUserEventData> {
+    const val = await this.getEventData(ON_SET_USER_TRACK_TYPE, {
```

Comment:
do we need this extra val const?

Reply:
No, we do not, it was a leftover of a test I was doing and I forgot to rollback. Reverted, thanks!